Us16 backend ingredientslogic3

akkuebler opened the pull request:
Hi Team, 
as its quite a big push including important logic, please find below comments/explanations. Content-wise its the same as alreday presented in yesterdays meeting (and the one on Friday). All is tested via Postman. As discussed, backend tests are missing and will be done next.

Step-by-step-approach:
1.) Frontend gets LIST of ingredients based on provided string (e.g. "ap" gives apple, apple sauce etc) to avoid spelling mistakes and enhances working with the external API. Rn they get 10 names send over (can be adjusted). Refer to postman for details
2.) Then frontend selects one of the provided names. They do it for every single "row" in the ingredient table-like structure. Saved in frontend for this moment and only after clicking in "submit ingredients" style button it is send to backend as a LIST full of objects. Backend needs sth like this  [{ "name": "fish" }, { "name": "beer"  }]
3.) User entity saves OBJECTS of ingredinets, not inly a list/set --> thats why I implemented a ManyToMany relationship
4.) ManyToMany also reflected in Ingredienst Class --> with Set<User> thingi
This is all for US 16 backend!

Next Steps: Add sth like "final User ID" in ingredients entity bc for group logic when we imagine that >1 contribute the same product, we need to filter only 1 person who in the end provides it (bc we do NOT deal with amounts/servings of product in this app). But that is US 17ff

Question to backend: 
PUT /user/{userId}/ingredients should have a 409 error mapping. Pls also see REST specification table in teams. Is this needed? Wouldn't it be nice to have a chance to update it? Do we have this usecase in future? Want to add it? 

As mentioned, my personal next todos are the tests for backend and cleaning up the classes.
Cheers, Anki

wlany reviewed (changes requested):
looks amazing! I also checked with postman :)

imo it makes sense to strike the 409 mapping as you suggested

akkuebler commented:
added all the tests now! 
Plus: changed the naming of the Ingredient method. Now its AddUserIngredient instead of UpdateUserIngredient bc its adding them to the repo, not updating completely. This should be ok imo as this is atm a one time thing (users fills out ingredient once) and after a group is finished all the ingredients associated with the user should be deleted anyway. 

To the tests: I hardly made them pass after some hours of trying. Thats why they might not be refactored nicely. But the should cover all possiblities, also the external API is tested
